### ISSUE TITLE
Resources<T> should be mapped to Collection<T>

### ISSUE THREAD

**Issue by Bert-R** [springfox/springfox#1025]
This issue is actually a rephrase of #1018, which was formulated as a question.

I'm having a method that returns `HttpEntity<Resources<GoalResource>>`. In SpringFox 2.2.2, this results in JSON that reflects the `Resources` class, while it should actually reflect a collection of `GoalResource`. On top of that, I want to map `GoalResource` to `GoalDTO`.

Here are more details:

This is the method that I'm trying to map, with the related classes:
```java
import org.springframework.hateoas.Resources;
import org.springframework.http.HttpEntity;

@RequestMapping(method = RequestMethod.GET)
@ResponseBody
public HttpEntity<Resources<GoalResource>> getAllGoals() {
	return createOKResponse(goalService.getAllGoals());
}

public static class GoalResource extends Resource<GoalDTO> {
	public GoalResource(GoalDTO goal) {
		super(goal);
	}
}
```

The controller class is available here: https://github.com/yonadev/yona-server/blob/master/src/main/java/nu/yona/server/goals/rest/GoalController.java

This is the mapping:
```java
AlternateTypeRule goalResourcesRule = newRule(
		typeResolver.resolve(HttpEntity.class, typeResolver.resolve(Resources.class, GoalResource.class)),
		typeResolver.resolve(List.class, GoalDTO.class));

.alternateTypeRules(goalResourcesRule)
```

The full docket is available here: https://github.com/yonadev/yona-server/blob/master/src/main/java/nu/yona/server/YonaServerApplication.java

In Swagger, I see this JSON:
```javascript
{
  "content": {
    "empty": true
  },
  "links": [
    {
      "href": "string",
      "rel": "string",
      "templated": true
    }
  ]
}
```

This corresponds with the `Resources` class, which has method `getContent` returning a `Collection`. Based on this, I conclude the mapping does not occur.

This is the JSON that I practically use and want to document:
```javascript
{
    "_links":{
        "self":{
            "href":"http://localhost:8080/goals/"
        }
    },
    "_embedded":{
        "Goals":[
            {
                "name":"g1",
                "categories":[
                    "a",
                    "b"
                ],
                "_links":{
                    "self":{
                        "href":"http://localhost:8080/goals/15c71c3b-94de-41db-a22b-0ffdfd8ecee5"
                    }
                }
            },
        ]
    }
}
```

Ideally, I shouldn't need a mapping for `Resources`, just like I don't need one for `Resource` anymore after upgrading to 2.2.2.

**Comment by dilipkrish** [comment]
@Bert-R that was a mass closing of all issues tagged as a question. May be I should put more rigor around the tagging :) ... in any case this is super useful!

**Comment by dilipkrish** [comment]
Just a quick suspicion that I'd like to throw out there. I think this :point_down: rule is taking precedence

```java
rules.add(newRule(typeResolver.resolve(HttpEntity.class, WildcardType.class),
            typeResolver.resolve(WildcardType.class)));
```

**Comment by Bert-R** [comment]
> that was a mass closing of all issues tagged as a question, in prep for 2.3.
No problem. This was the first GitHub issue I ever logged, so I thought I did something wrong.

> I think this :point_down: rule is taking precedence
That rule is inside SpringFox, right?
If you could add a rule in SpringFox that maps `Resources<T>` onto `Collection<T>`, my issue would be gone. Is that an option?

**Comment by Bert-R** [comment]
Given your suspicion, I tried excluding the `HttpEntity` from the rule, like this:

```java
AlternateTypeRule goalResourcesRule = newRule(
        Resources.class, GoalResource.class),
        typeResolver.resolve(List.class, GoalDTO.class));
```

But that doesn't help. Then the methods returning `HttpEntity<Resources<GoalResource>>` are not included in the Swagger doc anymore.

**Comment by dilipkrish** [comment]
The issue is really that the default rules should be lowest precedence... ill try to fix it such that your rule should just work after the change.

**Comment by Bert-R** [comment]
That makes sense indeed. Thanks in advance.

**Comment by dilipkrish** [comment]
This should be fixed in the next release (2.2.3-SNAPSHOT). Closing this issue of now. 

Regarding the documentation not looking like the hal media type, its because the object mapper that is being used doesn't have the rules that apply to the `_halObjectMapper` i.e. its missing the mixin configuration.

**Comment by Bert-R** [comment]
I've quickly tested this and indeed my rules are now applied with the 2.2.3 snapshot. Thanks!

Your second paragraph sounds like something I need to do. Do you have any pointers to documentation explaining what I should do?
